services/motion/builtin/move_request.go

```diff
@@ -85,12 +90,89 @@ func (mr *moveRequest) deviatedFromPlan(ctx context.Context, waypoints [][]refer
-	// TODO(RSDK-4507): implement this function
+	var plan motionplan.Plan
+	// We only care to check against waypoints we have not reached yet.
+	for _, inputs := range waypoints[waypointIndex:] {
+		input := make(map[string][]referenceframe.Input)
+		input[mr.kinematicBase.Name().Name] = inputs
+		plan = append(plan, input)
+	}
+	for _, cvMap := range mr.cameraVisionPairs {
+		for camName, visSrvc := range cvMap {
+			// get detections from vision service
+			detections, err := visSrvc.GetObjectPointClouds(ctx, camName.Name, nil)
+			if err != nil && strings.Contains(err.Error(), "does not implement a 3D segmenter") {
+				mr.planRequest.Logger.Debugf("cannot call GetObjectPointClouds on %q as it does not implement a 3D segmenter", visSrvc.Name())
+				break
+			} else if err != nil {
+				return false, err
+			}
+
+			// Note: detections are initially observed from the camera frame but must be transformed to be in
+			// world frame. We cannot use the inputs of the base to transform the detections since they are relative
+
+			// get the current position of the base which we will use to transform the detection into world coordinates
+			currentPosition, err := mr.kinematicBase.CurrentPosition(ctx)
+			if err != nil {
+				return false, err
+			}
+			// get transform of camera to kinematic base origin
+			kinBaseOrigin := referenceframe.NewPoseInFrame(mr.kinematicBase.Name().ShortName(), spatialmath.NewZeroPose())
+			cameraToBase, err := mr.frameSystemService.TransformPose(ctx, kinBaseOrigin, camName.ShortName(), nil)
+			if err != nil {
+				// here we make the assumption the movement sensor is coincident with the base
+				cameraToBase = kinBaseOrigin
+			}
+
+			// where the camera is in world coordinates
+			transformBy := spatialmath.Compose(currentPosition.Pose(), cameraToBase.Pose())
+
+			// Any obstacles specified by the worldstate of the moveRequest will also re-detected here.
+			// There is no need to append the new detections to the existing worldstate.
+			// We can safely build from scratch without excluding any valuable information.
+			geoms := []spatialmath.Geometry{}
+			for _, detection := range detections {
+				geometry := detection.Geometry.Transform(transformBy)
+				geometry.SetLabel("transient" + detection.Geometry.Label())
+				geoms = append(geoms, geometry)
+			}
+			gifs := []*referenceframe.GeometriesInFrame{referenceframe.NewGeometriesInFrame(referenceframe.World, geoms)}
+			worldState, err := referenceframe.NewWorldState(gifs, nil)
+			if err != nil {
+				return false, err
+			}
+
+			currentInputs, err := mr.kinematicBase.CurrentInputs(ctx)
+			if err != nil {
+				return false, err
+			}
+
+			// get the pose difference between where the robot is versus where it ought to be.
+			errorState, err := mr.kinematicBase.ErrorState(ctx, waypoints, waypointIndex)
+			if err != nil {
+				return false, err
+			}
+
+			if err := motionplan.CheckPlan(
```

Comment:
This is calling `CheckPlan` once per unique camera/vizService pair? How expensive is that?

Reply:
> This is calling CheckPlan once per unique camera/vizService pair?

yep

> How expensive is that?

for `TestReplanning`, running `CheckPlan` from `ensure no replan from discovered obstacles` took at most `11.612875ms` and at minimum `83.375µs`

for `ensure replan due to obstacle collision` took at most `110.667µs`

Reply:
Is there a reason this doesn't accumulate all geometries and call CheckPlan just once?

Reply:
I think its probably more efficient to batch these into one call.  Why do you feel that it is better to do it in two calls if they have diferent latencies? I am not following that logic

Reply:
suppose we have two obstacle detectors named `A` and `B` which are configured to poll for detections every 5 seconds.

Suppose that detector `A` takes 2 seconds to return, while `B` takes 4 seconds to return.

It is important to note that every time we get a detection back, we must convert the position of the detection to be in world coordinates since we cannot rely on the `ObstaclesInWorldFrame` method because the inputs which we would use to transform into the world frame are relative.

If we were to batch the detections so that we only call `CheckPlan` once, that would mean the `worldstate` we pass in as an argument must have its obstacles already in the world frame.

By the time we call `CheckPlan` on the batch of detections, the robot will be in a new place compared to where is was when `A` returned. I imagine this may introduce some error.

Further, in the best case scenario, the position and dimensions of detections returned by `A` should match the position and dimensions of detections returned by `B`. If that holds, in my opinion it would be best to run `CheckPlan` as soon as possible so that we return a possible collision sooner.

Reply:
This makes sense to me.